compaction/compaction_manager.cc

```diff
@@ -399,25 +401,116 @@ compaction_manager::run_with_compaction_disabled(replica::table* t, std::functio
     co_return;
 }
 
+std::string_view compaction_manager::task::to_string(state s) {
+    switch (s) {
+    case state::none: return "none";
+    case state::pending: return "pending";
+    case state::active: return "active";
+    case state::done: return "done";
+    case state::postponed: return "postponed";
+    case state::failed: return "failed";
+    }
+    __builtin_unreachable();
+}
+
+std::ostream& operator<<(std::ostream& os, compaction_manager::task::state s) {
+    return os << compaction_manager::task::to_string(s);
+}
+
 std::ostream& operator<<(std::ostream& os, const compaction_manager::task& task) {
     return os << task.describe();
 }
 
 std::string compaction_manager::task::describe() const {
-    auto* t = compacting_table();
+    auto* t = _compacting_table;
     auto s = t->schema();
     return fmt::format("{} task {} for table {}.{} [{}]", _description, fmt::ptr(this), s->ks_name(), s->cf_name(), fmt::ptr(t));
 }
 
+compaction_manager::task::~task() {
+    switch_state(state::none);
+}
+
+compaction_manager::sstables_task::~sstables_task() {
+    _cm._stats.pending_tasks -= _sstables.size() - (_state == state::pending);
+}
+
+future<> compaction_manager::task::run() noexcept {
+    try {
+        _compaction_done = do_run();
+        return compaction_done();
+    } catch (...) {
+        return current_exception_as_future();
+    }
+}
+
+compaction_manager::task::state compaction_manager::task::switch_state(state new_state) {
+    auto old_state = std::exchange(_state, new_state);
+    switch (old_state) {
+    case state::none:
+    case state::done:
+    case state::postponed:
+    case state::failed:
+        break;
+    case state::pending:
+        --_cm._stats.pending_tasks;
+        break;
+    case state::active:
+        --_cm._stats.active_tasks;
+        break;
+    }
+    switch (new_state) {
+    case state::none:
+    case state::postponed:
+    case state::failed:
+        break;
+    case state::pending:
+        ++_cm._stats.pending_tasks;
+        break;
+    case state::active:
+        ++_cm._stats.active_tasks;
+        break;
+    case state::done:
+        ++_cm._stats.completed_tasks;
+        break;
+    }
+    cmlog.debug("{}: switch_state: {} -> {}: pending={} active={} done={} errors={}", *this, _state, new_state,
```

Comment:
_state refers to new_state, after std::exchange(...) above, so we have to use old_state instead for the old state.

Reply:
True.  I moved the printout from the beginning of the function to the end, and missed that.  Will fix.